File: Services/AuditService.swift
import Foundation
import FirebaseFirestore
import FirebaseAuth
import CryptoKit

/// Service d'audit et de logs sécurisés pour traçabilité
class AuditService {
    
    static let shared = AuditService()
    private let db = Firestore.firestore()
    
    private init() {}
    
    // MARK: - Types d'événements
    
    enum AuditEventType: String, CaseIterable {
        case partnerConnection = "partner_connection"
        case partnerDisconnection = "partner_disconnection"
        case accessAttempt = "access_attempt"
        case accessDenied = "access_denied"
        case dataEncryption = "data_encryption"
        case dataDecryption = "data_decryption"
        case loginSuccess = "login_success"
        case loginFailure = "login_failure"
        case sensitiveDataAccess = "sensitive_data_access"
        case securityViolation = "security_violation"
    }
    
    // MARK: - Structure des événements
    
    struct AuditEvent {
        let type: AuditEventType
        let userId: String?
        let details: [String: Any]
        let timestamp: Date
        let ipAddress: String?
        let userAgent: String?
        let severity: Severity
        
        enum Severity: String {
            case low = "low"
            case medium = "medium"
            case high = "high"
            case critical = "critical"
        }
    }
    
    // MARK: - Logging des événements
    
    /// Logger un événement d'audit
    func logEvent(_ event: AuditEvent) {
        // Log local pour debug
        print("🔍 AuditService: \(event.type.rawValue) - Severity: \(event.severity.rawValue)")
        
        // Préparer les données pour Firestore
        var eventData: [String: Any] = [
            "type": event.type.rawValue,
            "timestamp": Timestamp(date: event.timestamp),
            "severity": event.severity.rawValue,
            "details": event.details
        ]
        
        // Ajouter userId de manière sécurisée
        if let userId = event.userId {
            // En production, hasher l'ID pour la confidentialité
            eventData["userId"] = hashUserId(userId)
            eventData["hashedUser"] = true
        }
        
        // Ajouter métadonnées techniques si disponibles
        if let ipAddress = event.ipAddress {
            eventData["ipAddress"] = ipAddress
        }
        if let userAgent = event.userAgent {
            eventData["userAgent"] = userAgent
        }
        
        // Sauvegarder dans Firestore
        db.collection("audit_events").addDocument(data: eventData) { error in
            if let error = error {
                print("❌ AuditService: Erreur sauvegarde audit - \(error)")
            } else {
                print("✅ AuditService: Événement d'audit sauvegardé")
            }
        }
    }
    
    // MARK: - Méthodes spécialisées
    
    /// Logger une connexion partenaire
    func logPartnerConnection(userId: String, partnerId: String, method: String) {
        let event = AuditEvent(
            type: .partnerConnection,
            userId: userId,
            details: [
                "partnerId": hashUserId(partnerId),
                "connectionMethod": method,
                "action": "successful_connection"
            ],
            timestamp: Date(),
            ipAddress: nil,
            userAgent: nil,
            severity: .medium
        )
        logEvent(event)
    }
    
    /// Logger une déconnexion partenaire
    func logPartnerDisconnection(userId: String, partnerId: String, reason: String) {
        let event = AuditEvent(
            type: .partnerDisconnection,
            userId: userId,
            details: [
                "partnerId": hashUserId(partnerId),
                "disconnectionReason": reason,
                "action": "partner_disconnected"
            ],
            timestamp: Date(),
            ipAddress: nil,
            userAgent: nil,
            severity: .medium
        )
        logEvent(event)
    }
    
    /// Logger un accès refusé
    func logAccessDenied(userId: String?, resource: String, reason: String) {
        let event = AuditEvent(
            type: .accessDenied,
            userId: userId,
            details: [
                "resource": resource,
                "denialReason": reason,
                "action": "access_blocked"
            ],
            timestamp: Date(),
            ipAddress: nil,
            userAgent: nil,
            severity: .high
        )
        logEvent(event)
    }
    
    /// Logger un accès aux données sensibles
    func logSensitiveDataAccess(userId: String, dataType: String, action: String) {
        let event = AuditEvent(
            type: .sensitiveDataAccess,
            userId: userId,
            details: [
                "dataType": dataType,
                "accessAction": action,
                "timestamp": ISO8601DateFormatter().string(from: Date())
            ],
            timestamp: Date(),
            ipAddress: nil,
            userAgent: nil,
            severity: .medium
        )
        logEvent(event)
    }
    
    /// Logger une violation de sécurité
    func logSecurityViolation(userId: String?, violationType: String, details: [String: Any]) {
        let event = AuditEvent(
            type: .securityViolation,
            userId: userId,
            details: details.merging([
                "violationType": violationType,
                "severity": "critical"
            ]) { (_, new) in new },
            timestamp: Date(),
            ipAddress: nil,
            userAgent: nil,
            severity: .critical
        )
        logEvent(event)
    }
    
    // MARK: - Utilitaires de sécurité
    
    /// Hasher un userId pour la confidentialité
    private func hashUserId(_ userId: String) -> String {
        // Utiliser SHA-256 pour hasher l'ID utilisateur
        let data = Data(userId.utf8)
        let hashed = SHA256.hash(data: data)
        return hashed.compactMap { String(format: "%02x", $0) }.joined()
    }
    
    /// Obtenir le userId actuel de manière sécurisée
    private func getCurrentUserId() -> String? {
        return Auth.auth().currentUser?.uid
    }
    
    // MARK: - Monitoring et alertes
    
    /// Vérifier les événements suspects récents
    func checkSuspiciousActivity(for userId: String, completion: @escaping (Bool) -> Void) {
        let oneDayAgo = Date().addingTimeInterval(-24 * 60 * 60)
        
        db.collection("audit_events")
            .whereField("userId", isEqualTo: hashUserId(userId))
            .whereField("severity", isEqualTo: "high")
            .whereField("timestamp", isGreaterThan: Timestamp(date: oneDayAgo))
            .getDocuments { snapshot, error in
                if let error = error {
                    print("❌ AuditService: Erreur vérification activité - \(error)")
                    completion(false)
                    return
                }
                
                let suspiciousCount = snapshot?.documents.count ?? 0
                let isSuspicious = suspiciousCount > 5 // Plus de 5 événements à risque en 24h
                
                if isSuspicious {
                    print("⚠️ AuditService: Activité suspecte détectée pour utilisateur")
                }
                
                completion(isSuspicious)
            }
    }
}

// MARK: - Extensions pour faciliter l'utilisation

extension AuditService {
    
    /// Logger rapidement un accès aux données de géolocalisation
    func logLocationAccess(action: String) {
        guard let userId = getCurrentUserId() else { return }
        logSensitiveDataAccess(userId: userId, dataType: "geolocation", action: action)
    }
    
    /// Logger rapidement un accès aux photos
    func logPhotoAccess(action: String) {
        guard let userId = getCurrentUserId() else { return }
        logSensitiveDataAccess(userId: userId, dataType: "profile_photo", action: action)
    }
    
    /// Logger rapidement un chiffrement/déchiffrement
    func logEncryptionActivity(type: String, success: Bool) {
        guard let userId = getCurrentUserId() else { return }
        let eventType: AuditEventType = type.contains("encrypt") ? .dataEncryption : .dataDecryption
        let severity: AuditEvent.Severity = success ? .low : .medium
        
        let event = AuditEvent(
            type: eventType,
            userId: userId,
            details: [
                "encryptionType": type,
                "success": success,
                "timestamp": ISO8601DateFormatter().string(from: Date())
            ],
            timestamp: Date(),
            ipAddress: nil,
            userAgent: nil,
            severity: severity
        )
        logEvent(event)
    }
}

File: Services/AuditService.kt
@file:Suppress("unused")
package com.love2love.security

import android.util.Log
import com.google.firebase.Timestamp
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.FirebaseFirestore
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.tasks.await
import kotlinx.coroutines.withContext
import java.security.MessageDigest
import java.util.Date

/**
 * Service d'audit et de logs sécurisés pour Android/Kotlin (Firestore + Firebase Auth)
 *\n * Remarques localisation : ce service ne contient pas de chaînes UI. Pour afficher un texte localisé
 * dans votre application, utilisez le mécanisme Android standard :
 *   context.getString(R.string.votre_cle)
 * ou en Compose :
 *   stringResource(id = R.string.votre_cle)
 */
object AuditService {

    private const val TAG = "AuditService"
    private val db: FirebaseFirestore by lazy { FirebaseFirestore.getInstance() }
    private val auth: FirebaseAuth by lazy { FirebaseAuth.getInstance() }

    // --- Types d'événements ---
    enum class AuditEventType(val raw: String) {
        PartnerConnection("partner_connection"),
        PartnerDisconnection("partner_disconnection"),
        AccessAttempt("access_attempt"),
        AccessDenied("access_denied"),
        DataEncryption("data_encryption"),
        DataDecryption("data_decryption"),
        LoginSuccess("login_success"),
        LoginFailure("login_failure"),
        SensitiveDataAccess("sensitive_data_access"),
        SecurityViolation("security_violation")
    }

    enum class Severity(val raw: String) { LOW("low"), MEDIUM("medium"), HIGH("high"), CRITICAL("critical") }

    data class AuditEvent(
        val type: AuditEventType,
        val userId: String?,
        val details: Map<String, Any?> = emptyMap(),
        val timestamp: Date = Date(),
        val ipAddress: String? = null,
        val userAgent: String? = null,
        val severity: Severity
    )

    /** Logger un événement d'audit */
    fun logEvent(event: AuditEvent) {
        // Log local pour debug
        Log.d(TAG, "${event.type.raw} - severity=${event.severity.raw}")

        // Préparer les données pour Firestore (sans valeurs nulles)
        val eventData = mutableMapOf<String, Any>(
            "type" to event.type.raw,
            "timestamp" to Timestamp(event.timestamp),
            "severity" to event.severity.raw,
            "details" to normalizeForFirestore(event.details)
        )

        // Ajouter userId de manière sécurisée (hash par défaut)
        event.userId?.let { uid ->
            eventData["userId"] = hashUserId(uid)
            eventData["hashedUser"] = true
        }
        event.ipAddress?.let { eventData["ipAddress"] = it }
        event.userAgent?.let { eventData["userAgent"] = it }

        db.collection("audit_events")
            .add(eventData)
            .addOnSuccessListener { Log.d(TAG, "Événement d'audit sauvegardé: ${it.id}") }
            .addOnFailureListener { e -> Log.e(TAG, "Erreur sauvegarde audit", e) }
    }

    // --- Méthodes spécialisées ---

    /** Logger une connexion partenaire */
    fun logPartnerConnection(userId: String, partnerId: String, method: String) {
        val event = AuditEvent(
            type = AuditEventType.PartnerConnection,
            userId = userId,
            details = mapOf(
                "partnerId" to hashUserId(partnerId),
                "connectionMethod" to method,
                "action" to "successful_connection"
            ),
            severity = Severity.MEDIUM
        )
        logEvent(event)
    }

    /** Logger une déconnexion partenaire */
    fun logPartnerDisconnection(userId: String, partnerId: String, reason: String) {
        val event = AuditEvent(
            type = AuditEventType.PartnerDisconnection,
            userId = userId,
            details = mapOf(
                "partnerId" to hashUserId(partnerId),
                "disconnectionReason" to reason,
                "action" to "partner_disconnected"
            ),
            severity = Severity.MEDIUM
        )
        logEvent(event)
    }

    /** Logger un accès refusé */
    fun logAccessDenied(userId: String?, resource: String, reason: String) {
        val event = AuditEvent(
            type = AuditEventType.AccessDenied,
            userId = userId,
            details = mapOf(
                "resource" to resource,
                "denialReason" to reason,
                "action" to "access_blocked"
            ),
            severity = Severity.HIGH
        )
        logEvent(event)
    }

    /** Logger un accès aux données sensibles */
    fun logSensitiveDataAccess(userId: String, dataType: String, action: String) {
        val event = AuditEvent(
            type = AuditEventType.SensitiveDataAccess,
            userId = userId,
            details = mapOf(
                "dataType" to dataType,
                "accessAction" to action,
                "timestamp" to Date().toInstant().toString()
            ),
            severity = Severity.MEDIUM
        )
        logEvent(event)
    }

    /** Logger une violation de sécurité */
    fun logSecurityViolation(userId: String?, violationType: String, details: Map<String, Any?>) {
        val merged = details.toMutableMap().apply {
            put("violationType", violationType)
            put("severity", Severity.CRITICAL.raw)
        }
        val event = AuditEvent(
            type = AuditEventType.SecurityViolation,
            userId = userId,
            details = merged,
            severity = Severity.CRITICAL
        )
        logEvent(event)
    }

    // --- Utilitaires de sécurité ---

    /** Hasher un userId pour la confidentialité (SHA-256, hex) */
    fun hashUserId(userId: String): String {
        val md = MessageDigest.getInstance("SHA-256")
        val digest = md.digest(userId.toByteArray())
        return buildString(digest.size * 2) {
            for (b in digest) append(String.format("%02x", b))
        }
    }

    /** Obtenir le userId courant via FirebaseAuth */
    fun getCurrentUserId(): String? = auth.currentUser?.uid

    // --- Monitoring & Alertes ---

    /**
     * Version coroutine (recommandée) : true si > 5 événements "high" en 24h
     */
    suspend fun checkSuspiciousActivity(userId: String): Boolean = withContext(Dispatchers.IO) {
        val oneDayAgo = Date(System.currentTimeMillis() - 24 * 60 * 60 * 1000L)
        val hashed = hashUserId(userId)
        try {
            val snap = db.collection("audit_events")
                .whereEqualTo("userId", hashed)
                .whereEqualTo("severity", Severity.HIGH.raw)
                .whereGreaterThan("timestamp", Timestamp(oneDayAgo))
                .get()
                .await()
            val count = snap.size()
            val suspicious = count > 5
            if (suspicious) Log.w(TAG, "Activité suspecte détectée pour l'utilisateur (hashed)")
            suspicious
        } catch (e: Exception) {
            Log.e(TAG, "Erreur vérification activité", e)
            false
        }
    }

    /**
     * Version callback (si vous n'utilisez pas les coroutines)
     */
    fun checkSuspiciousActivity(userId: String, onResult: (Boolean) -> Unit) {
        val oneDayAgo = Date(System.currentTimeMillis() - 24 * 60 * 60 * 1000L)
        val hashed = hashUserId(userId)
        db.collection("audit_events")
            .whereEqualTo("userId", hashed)
            .whereEqualTo("severity", Severity.HIGH.raw)
            .whereGreaterThan("timestamp", Timestamp(oneDayAgo))
            .get()
            .addOnSuccessListener { snap ->
                val suspicious = snap.size() > 5
                if (suspicious) Log.w(TAG, "Activité suspecte détectée pour l'utilisateur (hashed)")
                onResult(suspicious)
            }
            .addOnFailureListener { e ->
                Log.e(TAG, "Erreur vérification activité", e)
                onResult(false)
            }
    }

    // --- Extensions pratiques (mêmes intentions que l'extension Swift) ---

    /** Logger rapidement un accès aux données de géolocalisation */
    fun logLocationAccess(action: String) {
        val uid = getCurrentUserId() ?: return
        logSensitiveDataAccess(uid, dataType = "geolocation", action = action)
    }

    /** Logger rapidement un accès aux photos */
    fun logPhotoAccess(action: String) {
        val uid = getCurrentUserId() ?: return
        logSensitiveDataAccess(uid, dataType = "profile_photo", action = action)
    }

    /** Logger rapidement un chiffrement/déchiffrement */
    fun logEncryptionActivity(type: String, success: Boolean) {
        val uid = getCurrentUserId() ?: return
        val eventType = if (type.contains("encrypt", ignoreCase = true))
            AuditEventType.DataEncryption else AuditEventType.DataDecryption
        val severity = if (success) Severity.LOW else Severity.MEDIUM

        val event = AuditEvent(
            type = eventType,
            userId = uid,
            details = mapOf(
                "encryptionType" to type,
                "success" to success,
                "timestamp" to Date().toInstant().toString()
            ),
            severity = severity
        )
        logEvent(event)
    }

    // --- Normalisation Firestore ---

    /**
     * Firestore accepte String/Boolean/Number/Map/List/Timestamp/GeoPoint/DocumentReference.
     * Cette fonction normalise récursivement les Map/List et convertit Date en Timestamp.
     */
    private fun normalizeForFirestore(value: Any?): Any? = when (value) {
        null -> null
        is Date -> Timestamp(value)
        is Map<*, *> -> value.entries.associate { (k, v) ->
            k.toString() to normalizeForFirestore(v)
        }
        is List<*> -> value.map { normalizeForFirestore(it) }
        else -> value
    }
}
